Packet++/src/TcpReassembly.cpp

```diff
@@ -516,9 +516,12 @@ void TcpReassembly::checkOutOfOrderFragments(TcpReassemblyData* tcpReassemblyDat
 						}
 					}
 
-
-					// remove fragment from list
-					tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.erase(tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.begin() + index);
+					// fragment might be delete after calling m_OnMessageReadyCallback (eg. call TcpReassembly::closeConnection in m_OnMessageReadyCallback), so we need to check if it's already deleted 
+					if (tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.size() != 0)
```

Comment:
I think we should check if `tcpReassemblyData->twoSides[sideIndex].tcpFragmentList[tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.begin() + index]` exists, not checking the size?

Reply:
Do you mean `if (tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.begin() + index != tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.end()` ?

Reply:
I thought another problem: if we call reassemblePackage in m_OnMessageReadyCallback...... So maybe we should create a temporary fragment, pop the matehed fragment to the temporary one, send the temporary one to the m_OnMessageReadyCallback ?

Reply:
actually should be `if (tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.size() >  index)`, e.g. index=0, size=1; index=3, size=4

Reply:
> I thought another problem ...

could you try to write some tests for the cases that you mentioned? It will help us understand the issue easily.

Reply:
I think the `index` has been outdated due to the operation in `m_OnMessageReadyCallback`, the temporary fragment might be the best solution.

Reply:
I am not very familiar with this part of code, but it sounds correct.
But still, we will need some tests to verify our assumption.

Reply:
> > I thought another problem ...
> 
> could you try to write some tests for the cases that you mentioned? It will help us understand the issue easily.

This is the case which I face
```
void CaptureWorkerThread::tcpReassemblyMsgReadyCallback(const int8_t sideIndex, const pcpp::TcpStreamData& tcpData,
    void* userCookie)
{
    // extract the connection manager from the user cookie
    auto captureWorkerThread = static_cast<CaptureWorkerThread*>(userCookie);
    auto& connectionManager = captureWorkerThread->m_TcpConnectionManager;

    // check if this flow already appears in the connection manager. If not, igonre it
    auto flow = connectionManager.find(tcpData.getConnectionData().flowKey);
    if (flow == connectionManager.end()) {
        return;
        // connectionManager.insert(std::make_pair(tcpData.getConnectionData().flowKey, TcpConnectionData()));
        // flow = connectionManager.find(tcpData.getConnectionData().flowKey);
    }

    auto& connection = flow->second;

    if (connection.numOfDataPackets < 50) {
        connection.tcpData[connection.numOfDataPackets].side = sideIndex;
        connection.tcpData[connection.numOfDataPackets].data = tcpData.getData();
        connection.tcpData[connection.numOfDataPackets].length = tcpData.getDataLength();

        // count number of packets
        connection.numOfDataPackets += 1;

        // insert packets into database and close the connection after getting 50 packets in a flow
        if (connection.numOfDataPackets >= 50) {
            std::cout << "Inserting to Redis" << std::endl;
            captureWorkerThread->m_DataBase.insert(connection);
            connectionManager.erase(flow);
            captureWorkerThread->m_TcpReassembly.closeConnection(tcpData.getConnectionData().flowKey);
        }
    }
}
```

In this case, when `checkOutOfOrderFragments` calles `m_OnMessageReadyCallback` in the code above, the `m_OnMessageReadyCallback` calls `closeConnection`, and `closeConnection` calls `checkOutOfOrderFragments` again, the last `checkOutOfOrderFragments` clears the `tcpReassemblyData->twoSides[sideIndex].tcpFragmentList`, and causes error when the first `checkOutOfOrderFragments`  call `tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.erase(tcpReassemblyData->twoSides[sideIndex].tcpFragmentList.begin() + index);`

Reply:
Referring to your example, maybe the appropriate approach might be verifying whether the connection is closed subsequent to the callback invocation.

you can refer to some similar cases here, we have crazy manipulations in the callback in other places:
https://github.com/seladb/PcapPlusPlus/blob/5d217fff8a24d11a90d3cf088f85bd58d3b4125a/Tests/Pcap%2B%2BTest/Tests/LiveDeviceTests.cpp#L442-L450

Reply:
> Referring to your example, maybe the appropriate approach might be verifying whether the connection is closed subsequent to the callback invocation.
> 
> you can refer to some similar cases here, we have crazy manipulations in the callback in other places:
> 
> https://github.com/seladb/PcapPlusPlus/blob/5d217fff8a24d11a90d3cf088f85bd58d3b4125a/Tests/Pcap%2B%2BTest/Tests/LiveDeviceTests.cpp#L442-L450

I think verifying whether the connection is closed doesn't help when calling `reassemblePackage` in `m_OnMessageReadyCallback`, I will make a new commit to do with the temporary fragment soon.

Reply:
I believe that copying is not an optimal solution as it can reduce performance. The issue arises because `checkOutOfOrderFragments` is being called again while it is already in progress. To address this, consider introducing a flag that signals to the original `checkOutOfOrderFragments` whether it should continue execution after `m_OnMessageReadyCallback` returns. Alternatively, you can simply implement a check based on the size of the data as you currently did.

Reply:
> I believe that copying is not an optimal solution as it can reduce performance. The issue arises because `checkOutOfOrderFragments` is being called again while it is already in progress. To address this, consider introducing a flag that signals to the original `checkOutOfOrderFragments` whether it should continue execution after `m_OnMessageReadyCallback` returns. Alternatively, you can simply implement a check based on the size of the data as you currently did.

No copy, just wait a minute

Reply:
OK now, please check the new commit

Reply:
could you add a test in `Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp`?

we currently only have `tcpReassemblyMsgReadyCallback`, please have a test case that the connection is closed inside the callback.

Reply:
> could you add a test in `Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp`?
> 
> we currently only have `tcpReassemblyMsgReadyCallback`, please have a test case that the connection is closed inside the callback.

I replace the `int index` with `iterator`, and I remove the `fragment` from `fragmentList` before calling `m_OnMessageReadyCallback`  and freeing it, there is no need to check whether the connection is closed I think.

Reply:
We cannot ensure the logic correctness unless we have a test case to verify it. In this case, I would like to know if your patch can work for the aforementioned case.

Reply:
Sorry, I don't know write the test, but I can paste my code here

CaptureWorkerThread.h:
```
//
// Created by redbear on 24-3-7.
//

#ifndef CAPTURE_WORKER_H
#define CAPTURE_WORKER_H

#include <DnsLayer.h>
#include <DpdkDevice.h>
#include <DpdkDeviceList.h>
#include <IPLayer.h>
#include <TcpReassembly.h>

#include "DataBaes.h"
#include "TcpConnectionData.h"

#define MAX_RECEIVE_BURST 64

// typedef representing the connection manager
typedef std::unordered_map<uint32_t, TcpConnectionData> TcpConnectionManager;

class CaptureWorkerThread final : public pcpp::DpdkWorkerThread {
public:
    CaptureWorkerThread(uint32_t coreId, pcpp::DpdkDevice* device, int32_t rxQueueId);

    ~CaptureWorkerThread();

    bool run(uint32_t coreId) override;

    void stop() override;

    uint32_t getCoreId() const override;

private:
    static void handleDnsPacket(const pcpp::Packet& dnsPacket);

    /**
     * The callback being called by the TCP reassembly module whenever new data arrives on a certain connection
     */
    static void tcpReassemblyMsgReadyCallback(const int8_t sideIndex, const pcpp::TcpStreamData& tcpData,
        void* userCookie);

    /**
     * The callback being called by the TCP reassembly module whenever a new connection is found. This method adds the connection to the connection manager
     */
    static void tcpReassemblyConnectionStartCallback(const pcpp::ConnectionData& connectionData, void* userCookie);

    bool m_Stop = false;
    uint32_t m_CoreId;
    pcpp::DpdkDevice* p_Device;
    uint32_t m_RxQueueId;
    TcpConnectionManager m_TcpConnectionManager;
    pcpp::TcpReassemblyConfiguration m_TcpReassemblyConfiguration = pcpp::TcpReassemblyConfiguration(
        true, 5, 30, 50, true);
    pcpp::TcpReassembly m_TcpReassembly = pcpp::TcpReassembly(tcpReassemblyMsgReadyCallback, this,
        tcpReassemblyConnectionStartCallback,
        nullptr,
        // tcpReassemblyConnectionEndCallback,
        m_TcpReassemblyConfiguration);

    DataBase m_DataBase;
};

#endif // CAPTURE_WORKER_H

```

CaptureWorkerThread.cpp:
```
//
// Created by redbear on 24-3-7.
//

#include "CaptureWorkerThread.h"

#include "DataBaes.h"
#include "DnsBackTraceTable.h"

CaptureWorkerThread::CaptureWorkerThread(uint32_t coreId, pcpp::DpdkDevice* device, int32_t rxQueueId)
    : m_CoreId(coreId)
    , p_Device(device)
    , m_RxQueueId(rxQueueId)
{
}

CaptureWorkerThread::~CaptureWorkerThread()
{
}

bool CaptureWorkerThread::run(uint32_t coreId)
{
    pcpp::MBufRawPacket* mBufRawPackets[MAX_RECEIVE_BURST] = {};
    while (!m_Stop) {
        auto recievedPacketsCount = p_Device->receivePackets(mBufRawPackets, MAX_RECEIVE_BURST, m_RxQueueId);
        for (uint16_t i = 0; i < recievedPacketsCount; i += 1) {
            auto parsedPacket = pcpp::Packet(mBufRawPackets[i]);
            if (parsedPacket.isPacketOfType(pcpp::DNS))
                handleDnsPacket(parsedPacket);
            else {
                try {
                    m_TcpReassembly.reassemblePacket(parsedPacket);
                } catch (std::exception& e) {
                }
            }
        }
    }
    return true;
}

void CaptureWorkerThread::stop()
{
    m_Stop = true;
}

uint32_t CaptureWorkerThread::getCoreId() const
{
    return m_CoreId;
}

void CaptureWorkerThread::handleDnsPacket(const pcpp::Packet& dnsPacket)
{
    const auto dnsLayer = dnsPacket.getLayerOfType<pcpp::DnsLayer>();
    const auto ipLayer = dnsPacket.getLayerOfType<pcpp::IPLayer>();
    if (dnsLayer && ipLayer) {
        if (dnsLayer->getDnsHeader()->queryOrResponse == 1) {
            const auto timeStamp = dnsPacket.getRawPacketReadOnly()->getPacketTimeStamp().tv_sec;
            const auto query = dnsLayer->getFirstQuery();
            if (query) {
                const auto domain = query->getName();
                std::vector<std::string> resolvedIps;
                for (auto answer = dnsLayer->getFirstAnswer(); answer != nullptr;
                     answer = dnsLayer->getNextAnswer(answer)) {
                    if (answer->getDnsType() == pcpp::DNS_TYPE_A) {
                        auto resolvedIp = answer->getData().castAs<pcpp::IPv4DnsResourceData>()->getIpAddress().toString();
                        resolvedIps.push_back(resolvedIp);
                    } else if (answer->getDnsType() == pcpp::DNS_TYPE_AAAA) {
                        auto resolvedIp = answer->getData().castAs<pcpp::IPv6DnsResourceData>()->getIpAddress().toString();
                        resolvedIps.push_back(resolvedIp);
                    }
                }
                const auto queryIp = ipLayer->getDstIPAddress().toString();
                DnsBackTraceTable::updateTable(queryIp, resolvedIps, timeStamp, domain);
            }
        }
    }
}

void CaptureWorkerThread::tcpReassemblyMsgReadyCallback(const int8_t sideIndex, const pcpp::TcpStreamData& tcpData,
    void* userCookie)
{
    // extract the connection manager from the user cookie
    auto captureWorkerThread = static_cast<CaptureWorkerThread*>(userCookie);
    auto& connectionManager = captureWorkerThread->m_TcpConnectionManager;

    // check if this flow already appears in the connection manager. If not, igonre it
    auto flow = connectionManager.find(tcpData.getConnectionData().flowKey);
    if (flow == connectionManager.end()) {
        return;
        // connectionManager.insert(std::make_pair(tcpData.getConnectionData().flowKey, TcpConnectionData()));
        // flow = connectionManager.find(tcpData.getConnectionData().flowKey);
    }

    auto& connection = flow->second;

    if (connection.numOfDataPackets < 50) {
        connection.tcpData[connection.numOfDataPackets].side = sideIndex;
        connection.tcpData[connection.numOfDataPackets].data = tcpData.getData();
        connection.tcpData[connection.numOfDataPackets].length = tcpData.getDataLength();

        // count number of packets
        connection.numOfDataPackets += 1;

        if (connection.numOfDataPackets >= 50) {
            std::cout << "Inserting to Redis" << std::endl;
            captureWorkerThread->m_DataBase.insert(connection);
            connectionManager.erase(flow);
            captureWorkerThread->m_TcpReassembly.closeConnection(tcpData.getConnectionData().flowKey);
        }
    }
}

void CaptureWorkerThread::tcpReassemblyConnectionStartCallback(const pcpp::ConnectionData& connectionData,
    void* userCookie)
{
    // get a pointer to the connection manager
    auto captureWorkerThread = static_cast<CaptureWorkerThread*>(userCookie);
    auto& connectionManager = captureWorkerThread->m_TcpConnectionManager;
    // look for the connection in the connection manager
    auto connection = connectionManager.find(connectionData.flowKey);

    // assuming it's a new connection
    if (connection == connectionManager.end()) {
        // add it to the connection manager
        connectionManager.insert(std::make_pair(connectionData.flowKey, TcpConnectionData()));
    }
}

// void CaptureWorkerThread::tcpReassemblyConnectionEndCallback(const pcpp::ConnectionData& connectionData,
//     pcpp::TcpReassembly::ConnectionEndReason reason, void* userCookie)
// {
//     // get a pointer to the connection manager
//     auto captureWorkerThread = static_cast<CaptureWorkerThread*>(userCookie);
//     auto &connectionManager = captureWorkerThread->m_TcpConnectionManager;

//     // find the connection in the connection manager by the flow key
//     auto connection = connectionManager.find(connectionData.flowKey);

//     // connection wasn't found - shouldn't get here
//     if (connection == connectionManager.end())
//         return;

//     // remove the connection from the connection manager
//     connectionManager.erase(connection);
// }

```
At least in my case, this patch works fine

Reply:
> As for an additional test - if I understand correctly we want a test where we close a connection manually during an `OnTcpMessageReady` callback. It should be very complex to write. Let me know if you have an issue writing it, I might be able to help.

Yes, but I think maybe just modify current `TcpReassemblyTests.cpp` is enough.

Reply:
@MCredbear yes, we can add a test in `TcpReassemblyTests.cpp`. Do you think you can work on it?

Reply:
OK, I'll have a try.

Reply:
> we currently only have `tcpReassemblyMsgReadyCallback`, please have a test case that the connection is closed inside the callback.

As I mentioned above, you can see how we do for `tcpReassemblyMsgReadyCallback` in `TcpReassemblyTests.cpp`, and maybe have a callback `tcpReassemblyMsgCallbackConnectionClose`. You may also want to refer to `PcapPlusPlus/Tests/Pcap++Test/Tests/LiveDeviceTests.cpp` for more examples of testing callbacks.

Reply:
We currently have [`tcpReassemblyTest`](https://github.com/seladb/PcapPlusPlus/blob/5d217fff8a24d11a90d3cf088f85bd58d3b4125a/Tests/Pcap%2B%2BTest/Tests/TcpReassemblyTests.cpp#L177) that runs TCP reassembly and registers to the 3 callbacks (message ready, connections start, connection end). However, the [`tcpReassemblyMsgReadyCallback`](https://github.com/seladb/PcapPlusPlus/blob/5d217fff8a24d11a90d3cf088f85bd58d3b4125a/Tests/Pcap%2B%2BTest/Tests/TcpReassemblyTests.cpp#L98C13-L98C42) doesn't close the connection in the middle.

I think that modifying [`tcpReassemblyTest`](https://github.com/seladb/PcapPlusPlus/blob/5d217fff8a24d11a90d3cf088f85bd58d3b4125a/Tests/Pcap%2B%2BTest/Tests/TcpReassemblyTests.cpp#L177) to support this use case won't be easy. Hence, I'd recommend writing a new function, equivalent to [`tcpReassemblyTest`](https://github.com/seladb/PcapPlusPlus/blob/5d217fff8a24d11a90d3cf088f85bd58d3b4125a/Tests/Pcap%2B%2BTest/Tests/TcpReassemblyTests.cpp#L177) and a new "message ready callback" that closes the connection after it got X packets (while there are still out-of-ordered packets in the buffer).

@MCredbear if you need an existing scenario that has out-of-order packets you can use this test: https://github.com/seladb/PcapPlusPlus/blob/5d217fff8a24d11a90d3cf088f85bd58d3b4125a/Tests/Pcap%2B%2BTest/Tests/TcpReassemblyTests.cpp#L416